### ISSUE TITLE
Nested computed signals have glitchy behavior due to garbage collection of WeakRef dependencies

### ISSUE THREAD

**Issue by simeyla** [angular/angular#51055]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

There are many reasons you may end up with nested `computed()` signals, which could be deliberate or unintentionally. In the simplest form I'm talking about this, where a `computed` is created inside a computation and the value is accessed. Importantly, notice that `outerComputed` doesn't directly have a dependency on `counter`.

    const outerComputed = computed(() =>
    {
        const innerComputed = computed(() =>
        {
            return counter() % 5 == 0;
        });

        return innerComputed();
    });

Developers are most likely to encounter this without realizing if using something like [`selectSignal`](https://github.com/ngrx/platform/blob/47ef048213635c5395591d350189a932e7503d75/modules/component-store/src/component-store.ts#L328) from ngrx:

    const currentUserLastName = computed(() => store.selectSignal(s => s.users.filter(id => id == currentUserId()) )().lastName )

In the course of converting complex RxJS code (with nested `combineLatest` and `switchMap`) I have deliberately and unintentionally ended up with quite a lot of both of these in my codebase.

While they should probably generally be discouraged, an explicit 'inner computed' like the one above has the extremely useful characteristic that it can reduce unnecessary churn of the 'outer computed'. Of course that's the beauty of signals and is exactly the same as if  `innerComputed` was declared outside - but that may not always be possible.

In my example above `innerComputed` produces a `true` value every fifth time `counter` is incremented (then goes back to `false`). Since `innerComputed` is a direct dependency of `outerComputed` it is tracked internally via `producerAccessed()`. So when it changes, `outerComputed` will re-execute its computation (and yes that does recreate the innerComputed).

The problem I've come across is that since nothing ever saves a 'strong ref' to `innerComputed` it gets garbage collected sometime after the computation for `outerComputed` is run. That breaks the graph dependency from `outerComputed` which then stops working (at the whim of the garbage collector!)

Consequences / Observations: 
- If `outerComputed` had directly accessed `counter()` you would never notice a problem.
- If `innerComputed` was returned by `outerComputed` you would never notice a problem.
- If `outerComputed` has other dependencies, anytime one of those changes it will recreate and re-execute `innerComputed` so you perhaps would not notice a problem. This may be the hardest situation to debug!
- Overall this can lead to some incredibly glitchy behavior!

### Example
The Stackblitz example is a more comprehensive example of the above, with the addition of a second signal `animal()` that is a direct dependency of `outerComputed`. That can be used to trigger `outerComputed` to re-run its calculation (and recreate `innerComputed`) if it becomes broken. 

Here is a log of running this example:

- **Important:** Sometimes the garbage collector is much more aggressive than other times and it may take several attempts to observe this behavior.
- I am using Chrome on Windows 11. The garbage collector may behave differently elsewhere.
- Click 'Increment counter' 8 times and you should see something like the following:

> 14:36:22.299 main.ts:32 EFFECT: counter=0
> 14:36:22.299 main.ts:37 EFFECT: animal=cat
> 14:36:22.299 main.ts:49 executing computation. iteration 0
> 14:36:22.300 main.ts:61 EFFECT: outerComputed= {result: 'animal=cat, visible=true'}
> 14:36:23.380 main.ts:32 EFFECT: counter=1
> 14:36:23.380 main.ts:49 executing computation. iteration 1
> 14:36:23.380 main.ts:61 EFFECT: outerComputed= {result: 'animal=cat, visible=false'}
> 14:36:23.584 main.ts:32 EFFECT: counter=2
> 14:36:23.871 main.ts:32 EFFECT: counter=3
> 14:36:24.074 main.ts:32 EFFECT: counter=4
> 14:36:24.280 main.ts:32 EFFECT: counter=5
> 14:36:24.280 main.ts:49 executing computation. iteration 2
> 14:36:24.280 main.ts:61 EFFECT: outerComputed= {result: 'animal=cat, visible=true'}
> 14:36:24.504 main.ts:32 EFFECT: counter=6
> 14:36:24.504 main.ts:49 executing computation. iteration 3
> 14:36:24.504 main.ts:61 EFFECT: outerComputed= {result: 'animal=cat, visible=false'}
> 14:36:24.729 main.ts:32 EFFECT: counter=7
> 14:36:24.955 main.ts:32 EFFECT: counter=8

You can see the `effect` that logs `outerComputed()` was run only four times while `counter()` was incremented eight times.

Now wait a few seconds. The `innerComputed` is garbage collected (sometimes this takes longer of course). If you continue incrementing the counter neither `innerComputed` nor `outerComputed` will run anymore.

> 14:36:32.526 main.ts:32 EFFECT: counter=9
> 14:36:33.241 main.ts:32 EFFECT: counter=10
> 14:36:35.574 main.ts:32 EFFECT: counter=11
> 14:36:36.001 main.ts:32 EFFECT: counter=12
> 14:36:36.390 main.ts:32 EFFECT: counter=13
> 14:36:36.821 main.ts:32 EFFECT: counter=14
> 14:36:37.209 main.ts:32 EFFECT: counter=15

You can then toggle `animal` to restart things again.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/nested-computed-1

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.6
Node: 18.14.0
Package Manager: npm 9.3.1
OS: win32 x64

Angular: 16.1.5
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
```

### Anything else?

I really really don't want to lose the ability to create a `computed` inside another `computed`. For an inner calculation whose value rarely changes, it's a really nice way to reduce the burden on the primary computed and associated template. I have some cases where it works perfectly for what I need.

I also think if APIs are returning computed Signals it's going to be pretty much impossible to disallow it. 

I don't know enough about how the signals graphs works to know if it's even possible to get a good fix for (maybe a computed needs to keep stronger references to dependencies if it detects creation of inner computeds?).

**Comment by JeanMeche** [comment]
FYI, for the sake of reproductability it is possible to trigger the GC manually in the devTools ! 

![image](https://github.com/angular/angular/assets/1300985/13b29754-4958-46ec-88dd-458c715f066a)

**Comment by pkozlowski-opensource** [comment]
> I really really don't want to lose the ability to create a computed inside another computed. For an inner calculation whose value rarely changes, it's a really nice way to reduce the burden on the primary computed and associated template. I have some cases where it works perfectly for what I need.

I would love to hear more! 

The obvious source of trouble here is the inner computed creation inside a computed. Our intention (and intention of all signal libraries, in general) is that a computation function is a pure transformation, without side effects. This is not the case here as creating a new computed creates a node in a reactive graph. 

In any case, I can see 2 things to address here:
* our checks are insufficient - if our intention is to disallow creation of computed signals inside other computed signals we should do a better job here;
* the actual use-cases - we need to dig deeper into it since we need to make sure that the "well lit path" is the one that works and isn't introducing unnecessary friction. I can hear  "Developers are most likely to encounter this without realizing if using something like [selectSignal](https://github.com/ngrx/platform/blob/47ef048213635c5395591d350189a932e7503d75/modules/component-store/src/component-store.ts#L328) from ngrx:" (cc @markostanimirovic - those probably should be cached - although I'm unclear about the exact intention behind this API). 

tl;dr; is that intention here is _not_ to create computed inside a computed. The trivial change to the provided example would look like so: https://stackblitz.com/edit/nested-computed-1-3vwdvt?file=src%2Fmain.ts (pulling inner computed creation out of the outter one) - but this is the very thing you are trying to avoid. I'm trying to understand as of _why_ and I suspect that the part of the answer is in "I have some cases where it works perfectly for what I need." - we just need to dig deeper into it.

**Comment by simeyla** [comment]
I considered at one point titling this issue `Nested computed signals should be forbidden` but then I realized I couldn't survive if they were - even for some of the simplest of cases! I do fully understand all that you're saying though and there are clearly many DX and performance implications of any final decision. Wherever possible it's better to inform developers with warnings or errors about bad code they may have written, but as soon as you put up a solid wall it's going to break somebody's code.

### My most complicated use case

The code where I first came across this issue is quite complex and as I mentioned it was converted from RxJS (lots of `combineLatest` and `switchMap`). So I'm providing this overview mainly for curiosity - and then I'll move on to some more simplified and general thoughts.

I have components with dynamic config (from a JSON file) and that config can be updated (using Microsoft SignalR) while the page is running (via my own page editor). So if I update the design on my laptop my phone will instantly update by updating all the config signals. One way I've benefit from memoized nested `computed` signals is when accessing a single config property (maybe a boolean) which gets memoized in-place automatically without having to declare it 'outside'. This can help minimize recomputation or even refreshing of the whole component if other unconnected parts of the config change. It's just easier and cleaner to access the config in-place where I consume it, rather than 'outside' the computed. This notably never happens for the end user, their config never changes. So that's one way this behavior can benefit me, but not really useful as a general argument against nested computed signals because I could easily cache it 'outside' if I had to.

### Thoughts about APIs that use or return Signals

Consider a trivial API computation in a general third party / application specific service that consumes at least one global signal and returns a value. You first need to decide whether the API will return `Signal<T>` or just `T`. 

Return `Signal<T>`  (which is a `computed`)
- It's obvious that the value returned is reactive.
- The value returned can be cached easily (as a Signal).
- The value is already memoized.
- The value can be easily used in a page template.
- However the API call (that returns `computed`) CANNOT safely be made from inside another computed. (the crux of this issue)

Return just `T`
- The value is ready to use without dereferencing it.
- It's not necessarily obvious if the API call has consumed any signals (because it's just a boolean or an object). 
- The API call CAN safely be made inside a new computed. (because it can track all the dependencies)

You may even choose to offer both these options!
And projects like ngrx may have to end up doing so if nested computed were forbidden. 

I think on the whole API designers will choose to return a `Signal<T>` because it's more explicit and seems correct.

### Benefits of simply forbidding inner `computed`
- Users will quickly be able to see if they are recreating a signal on every iteration of their computation. That signal often can be moved outside to improve performance.
- Reduce graph churn, memory pressure and GC issues. This is especially an issue for a computation that gets run frequently or creates many inner `computed` signals.

### Disadvantages of forbidding inner computed
- Some code may end up harder to read if dependencies need to be declared 'outside'.
- Developers may end up having to provide `Signal<T>` and `T` return types from utility methods (as described above), adding complexity and confusion.
- It may be impossible to cleanly refactor code that creates a `computed` signal if the computation has both internal signal dependencies and requires inputs derived from application code signals. (I'll try to explain this below!)

I've no doubt that forbidding inner computed would suddenly break a lot of code (very curious about this one). I'm probably more convinced that it would fix a lot of silly mistakes. But I don't think it's feasible. 

### Use case for creating a Signal inside a computed
One example where I would have difficulty moving my dependencies 'outside' is for a list of dynamic components where the visibility of each is determined by a service. 

Consider an oversimplified 'visibility service' that determine if a page element (in a list of page elements) can be displayed or not. It might be based on screen width, global configuration, user role or AB testing settings.

The simplest implementation might look like this. 

```
class VisibilityService 
{
    // may depend on
    screenWidth: Signal<number>;
    isVideoEnabled: Signal<boolean>;
    userRole: Signal<string>;

    isVisible: (section: string) => boolean 
    isVisibleSignal: (section: string) => Signal<boolean> 
}
```

I've included methods with `boolean` and `Signal<boolean>` return types. They both internally run the same computation. 

eg. `isVisibleSignal(section: string) { return computed(() => this.isVisible(section)) }`

We then create a simple page model, passed to the template and displayed in an `*ngFor`.

```
// elements on the page
const page = signal( ['header', 'hero', 'video', 'footer'] );

const pageModel = computed(() => {

    return page().map(section =>
    {
        return {
            componentName: section,
            visible: visibilityService.isVisible(section)    
        }
    });
});
```

This is the simplest version, using `isVisible(string) => boolean`.  So that means no new `computed` is created by `visibilityService`, and any dependencies used by it are tracked by our `computed`. 

We could stop right here and probably never notice a problem :-) 

Now, if the calculation happens to depend on screen width you'd see terrible performance. So we need to switch to using the memoized value returned from `isVisibleSignal()` instead.

```
return page().map(section =>
{
    return {
        componentName: section,
        visible: visibilityService.isVisibleSignal(section)    
    }
});
```

This version creates new `computed` signals for each section, but never actually accesses them. Since it returns the signals they don't get garbage collected. 

Of course, now we need to use `visible()` inside the template (which is fine). Another benefit is that the `pageModel` computation (that creates all the visible signals) only gets executed if `page()` changes and not every time the visibility of an individual section changes.

If Angular had imposed a hard and fast rule `No computed can be created inside a computation` then this version would not be allowed (unfairly so I think). 

To trigger the GC issue we'd need to dereference these inner computed values:

```
return page().map(section =>
{
    return {
        componentName: section,
        visible: visibilityService.isVisibleSignal(section) ()
    }
});
```

Now we're back where we started (!) with an inner computed that suddenly stops working when GC kicks in. This version helps demonstrate its own inherent inefficiency since every visibility signal is recreated each time the visibility of any section changes. With a couple dozen sections could be very bad.

While we want to avoid creating new `computed` inside a computation as much as possible, we ultimately have a sort of chicken-and-egg problem:

* To get the list of page sections (which may change) we have to access `page()`
* To be reactive, the call to `page()` needs to be inside a `computed`
* To call the visibility service we need the value `section`
* But `section` only exists inside the `computed`
* So we must call the visibility service inside `computed`
* Now we have GC issues and graph side effects

Another approach might be to try and do it in two stages:

```
const page = signal( ['header', 'hero', 'video', 'footer'] );

// Create the page model with signals for visibility
const pageModel_intermediate = computed(() => {

    return page().map(section =>
    {
        return {
            componentName: section,

            // create the signal, but don't call it
            visible: visibilityService.isVisibleSignal(section)
        }
    });
});

// Step 2: Final model
const pageModel = computed(() => {

    const sections = pageModel_intermediate().map(section =>
    {
        return {
            ...section,
            visible: section.visible()
        }
    });

    // filter out only visible sections
    const visibleSections = sections.filter(s => s.visible);
    return visibleSections;
});
```

This actually works quite nicely: 

- Signals created by `visibilityService` are only created inside `pageModel_intermediate` when `page(s)` changes
- The final `pageModel` properly tracks all the `visible` signals as dependencies
- No GC issue
- No major graph issue. If `pages()` changes then all the previous signals are disconnected.

The problem here of course is:

- If Angular were to forbid `computed()` being called from a computation then we couldn't even do this
- Developers need to understand this whole issue to have got this far in the first place!

### Conclusions

- I don't think it's going to be possible for Angular to forbid the creation of nested/inner `computed` without a lot of pushback. 
- GC issue could be fixed if a `computed` computation were to track its own `ComputedImpl` dependencies and cache them with a strong reference.
- If Angular were to attempt to forbid inner `computed` solely to avoid the GC issue (unlikely) it would theoretically need to only forbid accessing newly created `computed` signals. But that sounds completely impractical and not worthy of further consideration.
- Whatever the outcome it may likely require an opt in option (similar to `allowSignalWrites`). This would help provide guidance and understanding of what's going on and prevent many unintentional nested `computed` signals.

**Comment by Harpush** [comment]
I think the issue here exists for rxjs too. When you have a template ngFor and a component signal and you need to compute something based on both - you are stuck.
You either need to prepare the data as viewmodel with the computed value as a property or just call a function that does the calculation without signals.
This issue usually happens in template and not outside - but it surely can.
I don't think nested signals are the solution here - but the fact remains that you need to do a lot of preparing if you want to compute something inside a loop without creating a different intermediate model.

**Comment by pkozlowski-opensource** [comment]
@simeyla thnx for sharing more info, this is helpful. I still think that we need to spend more time distilling the use-case / scenario - we will get there. 

IMO @Harpush comment hits the nail here and the way I start building the understanding of the use-case is that we've got 2 "levels" of computations:
* a "global" one from the config (info about sections visibility in your example);
* a user-specific one (represented by `isVideoEnabled`, `userRole` and `screenWidth` in your example);

At the end of the day we need a computed signal that combines those 2 pieces of info. Or, to put it differently - we need a computed per section.

What we are debating here are different patterns for creating this combined computed. I can see at least 3 different approaches (I'm sure there are more - and we will discover the best approach here!):

1.  try what @simeyla is persuading, but with a twist.
2. create a view model in TS code (@Harpush suggestion);
3. create a view model in a template (see below).

I think that we can make (1) work by skipping creating a computed signal and use a getter function: https://stackblitz.com/edit/stackblitz-starters-mapaoz?file=src%2Fmain.ts - creating a computed per-section is probably more trouble (code _and_ perf) than it is worth so this would be probably my preferred solution.

An alternative would be to return a _cached_ computed from `VisibilityService.isVisible` but this indeed requires that we allow creation of computed inside a computation function (cache here would allow us to keep references and lazy-create them).

But if we want to create those combined computed per section we could do it explicitly in the TS code (as suggested by @Harpush ).

With the signal-based components and the updated control flow we should be able to create those computed when declaring what goes into the context of a template created by ngFor. This will come with the full integration of signals into Angular.

As @Harpush I don't think that nested computed signals are the solution here but I don't want to make hasty decisions here as I can see how creating a _cached_ computed could be useful in _some_ cases. But, while being useful, it might have too many sharp edges to allow in practice.

I'm going to keep this issue open so others can share their approaches / thoughts. In the meantime I'm going to check why we don't report creation of computed in a computed - it should be _at least_ a warning.

**Comment by Harpush** [comment]
@pkozlowski-opensource out of curiosity - computed as part of control flow was mentioned in the RFC comments only - do you have a clear idea how it will work?

**Comment by mfp22** [comment]
What I would do is put a debounceTime on the rapidly changing signals and grab them in the main computed and just recompute the list every time. Make sure to use `trackBy`. 

However, I know SolidJS stores use nested signals, so in the long run I think something like nested signals could be possible.

As for the debounceTime, that would require using an `effect` to manage. This is something I'm putting in my next article:

> When you think you might need to write an effect for a signal, first try instead to
> 1. reorganize the code to use a `computed` 
> 2. use a custom signal operator
> 3. write a new custom signal operator
> 4. convert to an observable with `toObservable` and use an RxJS operator

So if you have time I'd create a `debounceTime` function that takes in a source signal, returns an output signal, and has an effect to manage the timeout for the debounce. Otherwise, use RxJS.

If I failed to understand the problem, can you please create a single code snippet that has all the elements of the problem? It's taking me a long time to read everything and remember all the details.

**Comment by simeyla** [comment]
Despite my very long previous answer (!) it's probably not too helpful to try to look for a complete solution to my specific case here. Not that all ideas or possibilities aren't appreciated and welcome &ndash; and being eagerly added to my signals toolkit. My code is much more complex than my above examples and I don't think it should be the primary focus here. I also definitely want to stay as synchonous as possible and solve this only with signals.

Fortunately I can trivially 'solve' my main issue by simply running my 'inner computation' every time. They use only pure signal reads so I don't break any 'signal rules' or need RxJS.  It's just a *tiny* bit less efficient.

### Questions

I think it's better going forward to focus on general questions such as:

- What warnings are given if you create a new `signal` / `computed` inside a `computed`.
- Is the warning an error, requiring a flag?
- What use cases might people have where they would run into this?
- What guidance or workarounds should Angular provide?
- Most importantly: How to handle the garbage collector issue. eg. What should happen if you create and use a memoized computation inside another computation that gets collected. Such behavior is all too [easy to trigger](https://github.com/ngrx/platform/issues/3955).

### Simplest use case

I've tried to distill the crux of what I am trying to do to the simplest possible use case:

- I have a factory that returns a memoized `Signal<Y>`, dependent on a single immutable input `X` and global signal `G`

   ```(x: X) => Signal<Y>```

- I want to create one of these with a value for `x` that comes from `source: Signal<X>`
- If `source` changes I need the memoized signal to update.
- Last week that sounded like a simple requirement (and is trivial with RxJS). 
- But you can't easily do it.

**Comment by markostanimirovic** [comment]
> The obvious source of trouble here is the inner computed creation inside a computed. Our intention (and intention of all signal libraries, in general) is that a computation function is a pure transformation, without side effects. This is not the case here as creating a new computed creates a node in a reactive graph.

💯 Nested computed signals don't seem like good practice to me and I agree that there should be at least a warning.

---

@simeyla Is there really a need for nested computed signals in your case? If you're using nested signals to improve performance via memoization, what do you think about creating a (pure) memoized function for nested calculation instead?

---

Regarding NgRx, we could cache signals created by the `selectSignal` method, but the issue with nested computed signals will be still present. So if we want to allow this 'feature' at all, I think it would be better to solve it on the Angular side.

**Comment by simeyla** [comment]
@markostanimirovic So I don't really set out to intentionally to 'create nested signals'. It's always as a side effect of trying to use a cached / memoized calculation.

I like the sound of `(pure) memoized function` but that's sort of what I thought I was doing! 

Can you give an example of what you mean by that?

**Comment by mfp22** [comment]
> ### Simplest use case
> I've tried to distill the crux of what I am trying to do to the simplest possible use case:
> 
> * I have a factory that returns a memoized `Signal<Y>`, dependent on a single immutable input `X` and global signal `G`
>   `(x: X) => Signal<Y>`
> * I want to create one of these with a value for `x` that comes from `source: Signal<X>`
> * If `source` changes I need the memoized signal to update.
> * Last week that sounded like a simple requirement (and is trivial with RxJS).
> * But you can't easily do it.

Thanks for the distilled description. 

Signals aren't lazy. The signals used in a component are supposed to exist throughout the lifecycle of the component. Currently you can pass an injector ref to an effect, but I don't think you can do that for signals. 

I don't know about about Angular's future, but `useState` in React, `useSelector$` in Qwik, `createSignal` in SolidJS and... I think I'm forgetting something, but all of those have strict rules around state/signal creation. At the current time, the right mental model is that signals are top-level in components. I still haven't really seen a reason the nested signals have to exist, although I don't doubt there could be reasons. 

So I would say either don't do what you're doing, or do it with RxJS, or do it with child components as inner-signal platforms. The number of signals depends on the output of another signal? Then you can render child components in an `NgFor` using the output of `Signal<X>`, then inside the child component you can create signals. This means signals are being created and destroyed according to the output of the first signal. What you actually render inside that child component can then depend on the results of the "inner" signal.

An issue would come up if you need to share these inner signals with multiple components. And you probably don't know up-front if you will eventually have to or not.

There are other weird patterns you could try if you're desperate. One would be creating placeholder signals and using effects to connect them. All the weird patters I'm thinking of are variations of this.

**Comment by markostanimirovic** [comment]
> I like the sound of (pure) memoized function but that's sort of what I thought I was doing!
>
> Can you give an example of what you mean by that?

I'd like to see the full example, but based on the code that is provided, you can do something like this:

```ts
const pageModel = computed(() => {
  const screenWidth = visibilityService.screenWidth();
  const isVideoEnabled = visibilityService.isVideoEnabled();
  const userRole = visibilityService.userRole();

  return page().map((section) => ({
    componentName: section,
    isVisible: isSectionVisible(section, screenWidth, isVideoEnabled, userRole),
  }));
});

// you can use `memoize` function from `lodash` or check another implementation
const isSectionVisible = memoize((
  section: string,
  screenWidth: number,
  isVideoEnabled: boolean,
  userRole: string,
) => {
  /* logic won't be re-executed when it's called again with the same input args */
});
```

I also like what Mike proposed - create `isVisible` signal in the child component:

```html
<!-- parent -->
<app-child *ngFor="let page of pages()" [page]="page"  />
```

```ts
// child
@Component(/* ... */)
export class ChildComponent {
  private visibilityService = inject(VisibilityService);

  page = signal<string | null>(null);
  isVisible = computed(() => {
    const page = this.page();
    const screenWidth = this.visibilityService.screenWidth();
    const isVideoEnabled = this.visibilityService.isVideoEnabled();
    const userRole = this.visibilityService.userRole();

    // check if the page is visible
  });

  // with signal inputs, we'll be able to remove this part
  @Input({ required: true, alias: 'page' })
  set pageInput(page: string) {
    this.page.set(page);
  }
}
```

**Comment by pkozlowski-opensource** [comment]
> I think it's better going forward to focus on general questions such as:
> * What warnings are given if you create a new signal / computed inside a computed.
> * Is the warning an error, requiring a flag?
> * What use cases might people have where they would run into this?
> * What guidance or workarounds should Angular provide?

Currently leaning towards disallowing this as a hard error. And gradually relax this error if important and legitimate use-cases become clear. We are still in the developer preview, dealing with the experimental APIs so while this change is disruptive, it is our only chance to put guide-rails in place.

> Most importantly: How to handle the garbage collector issue. eg. What should happen if you create and use a memoized computation inside another computation that gets collected. Such behavior is all too https://github.com/ngrx/platform/issues/3955.

Per above, I don't think it should be allowed. We will work with the NgRX team to make iron out the API and make sure that people are not bumping into it for common scenarios.

**Comment by simeyla** [comment]
@pkozlowski-opensource Thanks for addressing this. Agree with the change. 

It will be very interesting to see what use cases this might reveal - and help fix some bugs at the same time.

I see now `untracked` has two undocumented side effects, to allow signal writes (as utilized by async pipe) and also to circumvent this new check if absolutely necessary ;-)

**Comment by pkozlowski-opensource** [comment]
With #51226 merged and WeakRef gone we should be able to create `computed` inside a `computed` now. The change is available in [17.0.0-next.3](https://github.com/angular/angular/releases/tag/17.0.0-next.3)

It is still not a recommended approach but at least it should behave correctly.
